libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,478 @@
+        .primary(args_.x)
+        .emit(dh);
+      return false;
+    }
+    if (args_.resolution and not ty.kind().is_any<time_type, duration_type>()) {
+      diagnostic::warning("cannot group type `{}` with resolution", ty.kind())
+        .note("skipping invalid events")
+        .primary(args_.x)
+        .primary(args_.resolution->source)
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  auto validate_ytype(const type& ty, diagnostic_handler& dh) const -> bool {
+    if (not ty.kind()
+              .is_any<int64_type, uint64_type, double_type, duration_type,
+                      time_type>()) {
+      diagnostic::warning("Y-axis cannot have type `{}`", ty.kind())
+        .primary(args_.y)
+        .emit(dh);
+      return false;
+    }
+    return true;
+  }
+
+  // Modified from `floor()`
+  auto floor_xs(const series& xs) const -> series {
+    return match(
+      *xs.array,
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(i);
+          const auto count = std::abs(args_.resolution->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem == 0) {
+            check(b->Append(val));
+            continue;
+          }
+          const auto floor = val >= 0 ? -rem : rem - count;
+          check(b->Append(val + floor));
+        }
+        return {duration_type{}, finish(*b)};
+      },
+      [&](const arrow::TimestampArray& array) -> series {
+        auto opts = make_round_temporal_options(args_.resolution->inner);
+        return {time_type{},
+                check(arrow::compute::FloorTemporal(array, std::move(opts)))
+                  .array_as<arrow::TimestampArray>()};
+      },
+      [&](const auto&) -> series {
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.chart";
+  }
+
+  auto optimize(const expression& filter,
+                event_order) const -> optimize_result override {
+    auto expr = filter;
+    if (args_.from) {
+      expr = conjunction{args_.from->first, std::move(expr)};
+    }
+    if (args_.to) {
+      expr = conjunction{args_.to->first, std::move(expr)};
+    }
+    auto combined = normalize_and_validate(expr);
+    TENZIR_ASSERT(combined);
+    return {std::move(combined).value(), event_order::ordered, copy()};
+  }
+
+  friend auto inspect(auto& f, chart_operator2& x) -> bool {
+    return f.object(x)
+      .pretty_name("chart_operator2")
+      .fields(f.field("args_", x.args_));
+  }
+
+private:
+  chart_args args_;
+  // Using a `deque` to guarantee reference validity after growing
+  mutable std::deque<std::string> ys{"y"};
```

Comment:
The operator class should not have any mutable state.

Reply:
Why is that?